File: src/app/services/fraud-alert.service.ts
import { Injectable, signal, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, tap, catchError, of, map } from 'rxjs';
import { environment } from '../../environments/environment';
import { FraudAlert } from '../models/alert.model';
import { ApiResponse, PagedResult, FraudAlertDto } from '../models/api-response.model';
import { AlertStatus } from '../models/types.model';

export function mapDtoToAlert(dto: FraudAlertDto): FraudAlert {
  return {
    id: dto.alertCode || dto.alertId,
    transactionId: dto.transactionCode || dto.transactionId,
    customerId: dto.customerId,
    customerName: dto.customerName || 'Unknown Customer',
    amountInr: dto.amountInr,
    fraudProbability: dto.fraudProbability,
    riskTier: (dto.riskTier as any) || (dto.severity as any) || 'Medium',
    reason: dto.reason,
    status: (dto.status as any) || 'Open',
    assignedTo: dto.assignedTo,
    createdAt: dto.createdDateFormatted || new Date(dto.createdAt).toLocaleDateString(),
    createdTime: dto.createdTimeFormatted || new Date(dto.createdAt).toLocaleTimeString(),
    location: dto.location || 'Unknown Location',
    alertType: dto.alertType || 'Suspicious Activity'
  };
}

@Injectable({
  providedIn: 'root'
})
export class FraudAlertService {
  private readonly http = inject(HttpClient);

  // Reactive Signals
  alerts = signal<FraudAlert[]>([]);
  loading = signal<boolean>(false);
  error = signal<string | null>(null);
  totalCount = signal<number>(0);

  constructor() {
    this.loadAlerts();
  }

  loadAlerts(
    status?: string,
    severity?: string,
    page: number = 1,
    pageSize: number = 50
  ): Observable<FraudAlert[]> {
    this.loading.set(true);
    this.error.set(null);

    let params = new HttpParams()
      .set('page', page.toString())
      .set('pageSize', pageSize.toString());

    if (status && status !== 'All') params = params.set('status', status);
    if (severity && severity !== 'All') params = params.set('severity', severity);

    return this.http.get<ApiResponse<PagedResult<FraudAlertDto>>>(`${environment.apiUrl}/fraud-alerts`, { params }).pipe(
      map(res => {
        if (res && res.success && res.data && res.data.items) {
          const mapped = res.data.items.map(mapDtoToAlert);
          this.alerts.set(mapped);
          this.totalCount.set(res.data.totalCount);
          this.loading.set(false);
          return mapped;
        }
        this.loading.set(false);
        return [];
      }),
      catchError(err => {
        this.loading.set(false);
        this.error.set(err.message || 'Failed to load fraud alerts');
        return of(this.alerts());
      })
    );
  }

  getAlerts(): FraudAlert[] {
    return this.alerts();
  }

  getAlertById(id: string): FraudAlert | undefined {
    return this.alerts().find(a => a.id.toLowerCase() === id.toLowerCase() || (a as any).alertId === id);
  }

  resolveAlert(id: string, notes: string = 'Resolved by investigator review'): Observable<ApiResponse<FraudAlertDto>> {
    return this.http.patch<ApiResponse<FraudAlertDto>>(`${environment.apiUrl}/fraud-alerts/${id}/resolve`, {
      resolutionNotes: notes
    }).pipe(
      tap(() => {
        this.alerts.update(alerts =>
          alerts.map(a => (a.id.toLowerCase() === id.toLowerCase() ? { ...a, status: 'Resolved' as AlertStatus } : a))
        );
      })
    );
  }

  assignAlert(id: string, assignedUserId: string): Observable<ApiResponse<FraudAlertDto>> {
    return this.http.patch<ApiResponse<FraudAlertDto>>(`${environment.apiUrl}/fraud-alerts/${id}/assign`, {
      assignedToUserId: assignedUserId
    }).pipe(
      tap(() => {
        this.alerts.update(alerts =>
          alerts.map(a => (a.id.toLowerCase() === id.toLowerCase() ? { ...a, status: 'Investigating' as AlertStatus } : a))
        );
      })
    );
  }
}

File: src/backend/FraudGuard.Api/Services/FraudAlertService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using FraudGuard.Api.Data;
using FraudGuard.Api.DTOs.Common;
using FraudGuard.Api.DTOs.FraudAlerts;

namespace FraudGuard.Api.Services
{
    public interface IFraudAlertService
    {
        Task<PagedResult<FraudAlertDto>> GetAlertsAsync(string? status, string? severity, int page, int pageSize);
        Task<FraudAlertDto?> GetAlertByIdAsync(Guid id);
        Task<bool> ResolveAlertAsync(Guid id, string? notes);
        Task<bool> AssignAlertAsync(Guid id, Guid userId);
    }

    public class FraudAlertService : IFraudAlertService
    {
        private readonly FraudGuardDbContext _db;

        public FraudAlertService(FraudGuardDbContext db)
        {
            _db = db;
        }

        public async Task<PagedResult<FraudAlertDto>> GetAlertsAsync(string? status, string? severity, int page, int pageSize)
        {
            page = Math.Max(1, page);
            pageSize = Math.Clamp(pageSize, 1, 100);

            var query = _db.FraudAlerts
                .Include(fa => fa.Customer)
                .Include(fa => fa.Transaction)
                    .ThenInclude(t => t!.FraudPrediction)
                .Include(fa => fa.AssignedToUser)
                .AsNoTracking();

            if (!string.IsNullOrWhiteSpace(status) && !status.Equals("All", StringComparison.OrdinalIgnoreCase))
            {
                query = query.Where(fa => fa.Status == status);
            }

            if (!string.IsNullOrWhiteSpace(severity) && !severity.Equals("All", StringComparison.OrdinalIgnoreCase))
            {
                query = query.Where(fa => fa.Severity == severity);
            }

            var totalCount = await query.CountAsync();

            var items = await query
                .OrderByDescending(fa => fa.CreatedAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .Select(fa => new FraudAlertDto
                {
                    AlertId = fa.AlertId,
                    AlertCode = fa.AlertCode,
                    TransactionId = fa.TransactionId,
                    TransactionCode = fa.Transaction != null ? fa.Transaction.TransactionCode : "",
                    CustomerId = fa.CustomerId,
                    CustomerName = fa.Customer != null ? fa.Customer.FullName : "",
                    AmountInr = fa.Transaction != null ? fa.Transaction.AmountInr : 0,
                    FraudProbability = fa.Transaction != null && fa.Transaction.FraudPrediction != null ? fa.Transaction.FraudPrediction.FraudProbability : 0,
                    RiskTier = fa.Severity,
                    Severity = fa.Severity,
                    Reason = fa.Reason,
                    Status = fa.Status,
                    AssignedTo = fa.AssignedToUser != null ? fa.AssignedToUser.FullName : null,
                    Location = fa.Transaction != null ? fa.Transaction.City + ", " + fa.Transaction.Country : "",
                    AlertType = fa.AlertType,
                    CreatedAt = fa.CreatedAt,
                    CreatedDateFormatted = fa.CreatedAt.ToString("MMM dd, yyyy"),
                    CreatedTimeFormatted = fa.CreatedAt.ToString("HH:mm IST")
                })
                .ToListAsync();

            return new PagedResult<FraudAlertDto>
            {
                Items = items,
                TotalCount = totalCount,
                Page = page,
                PageSize = pageSize
            };
        }

        public async Task<FraudAlertDto?> GetAlertByIdAsync(Guid id)
        {
            var fa = await _db.FraudAlerts
                .Include(x => x.Customer)
                .Include(x => x.Transaction)
                    .ThenInclude(t => t!.FraudPrediction)
                .Include(x => x.AssignedToUser)
                .AsNoTracking()
                .FirstOrDefaultAsync(x => x.AlertId == id);

            if (fa == null) return null;

            return new FraudAlertDto
            {
                AlertId = fa.AlertId,
                AlertCode = fa.AlertCode,
                TransactionId = fa.TransactionId,
                TransactionCode = fa.Transaction != null ? fa.Transaction.TransactionCode : "",
                CustomerId = fa.CustomerId,
                CustomerName = fa.Customer != null ? fa.Customer.FullName : "",
                AmountInr = fa.Transaction != null ? fa.Transaction.AmountInr : 0,
                FraudProbability = fa.Transaction != null && fa.Transaction.FraudPrediction != null ? fa.Transaction.FraudPrediction.FraudProbability : 0,
                RiskTier = fa.Severity,
                Severity = fa.Severity,
                Reason = fa.Reason,
                Status = fa.Status,
                AssignedTo = fa.AssignedToUser != null ? fa.AssignedToUser.FullName : null,
                Location = fa.Transaction != null ? fa.Transaction.City + ", " + fa.Transaction.Country : "",
                AlertType = fa.AlertType,
                CreatedAt = fa.CreatedAt,
                CreatedDateFormatted = fa.CreatedAt.ToString("MMM dd, yyyy"),
                CreatedTimeFormatted = fa.CreatedAt.ToString("HH:mm IST")
            };
        }

        public async Task<bool> ResolveAlertAsync(Guid id, string? notes)
        {
            var alert = await _db.FraudAlerts.FindAsync(id);
            if (alert == null) return false;

            alert.Status = "Resolved";
            alert.ResolvedAt = DateTimeOffset.UtcNow;
            alert.UpdatedAt = DateTimeOffset.UtcNow;
            await _db.SaveChangesAsync();
            return true;
        }

        public async Task<bool> AssignAlertAsync(Guid id, Guid userId)
        {
            var alert = await _db.FraudAlerts.FindAsync(id);
            if (alert == null) return false;

            alert.AssignedToUserId = userId;
            alert.Status = "Investigating";
            alert.UpdatedAt = DateTimeOffset.UtcNow;
            await _db.SaveChangesAsync();
            return true;
        }
    }
}
